### ISSUE TITLE
blank messages can be sent through the Private Message (PM) system

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#4872]
The private message system should prevent a message being sent if nothing has been typed. It seems that players will accidentally click the message icon (possibly instead of the like button) and then accidentally click send instead of cancel. 

Clicking send on an empty message should close the message window without sending anything.

Note that this change should not affect gifting of gems - it should still be possible to send gems with no message.

**Comment by FriendlyPopo** [comment]
In a similar vein, the API also allows sending empty messages to group. Should this be in the same issue? I understand they're not really related in the code, but the fixes are likely to be similar.

**Comment by Alys** [comment]
Actually, that's a good point. If we made the API silently discard any empty messages, either PMs or group messages, then that's all that would be needed. We wouldn't have to change any of the website code.

**Comment by TheHollidayInn** [comment]
Should we not notify the user that the message is empty?

**Comment by FriendlyPopo** [comment]
@TheHollidayInn Agreed. From a UX perspective, it is indeed important to be clearer about system status than just allowing the post and then not posting it in the backend.

That being said, once again from a UX perspective, the best solution would be to prevent it all together: disable "Send" buttons in the UI until there is data in the textarea.

API should simply return the appropriate error.

**Comment by Alys** [comment]
The API could return an error, but on the front end there should be no notification to the user. If they click send without having entered a message, it's almost certainly because they meant to click cancel, and so giving them the cancel behaviour is appropriate.

**Comment by FriendlyPopo** [comment]
@Alys Unfortunately I'll have to disagree with that. We do not know that if users click send without having entered a message it's because they meant to click cancel. While it may seem logical, it's still speculation.

I believe there might be unexpected scenarios where sending a message is precisely what was intended. ie User had all text highlighted because they wanted to copy it beforehand, then accidentally deletes it on their way to the "send" button... I understand this is an edge case, but regardless: if a button says "send", it should only send.

I stick to the fact that "error prevention" > "unclear behavior".

**Comment by Alys** [comment]
If you wish. I still think canceling would be more often useful in practice but you are correct that it is not the function of the send button. I won't argue if you submit a PR for your idea. :)

**Comment by rdhammond** [comment]
I think (agree?) we need both: Something to disable the "Send" button website-side as a courtesy, and something in the API to block that invalid state from ever happening.

The former should be _fairly_ easy, just slap a $watch() or jquery.on() or some such to disable and enable the button. It's just a boolean check so it shouldn't introduce any drag :) If no one's got anything in the works, I can bang it out real quick.

**Comment by crookedneighbor** [comment]
There's an `ng-disabled` directive that will do the work of disabling the send button when the message is blank.

**Comment by rdhammond** [comment]
Neato. I am understanding correctly that I need to combine ng-change on the box with ng-disabled on the button for full handling?

**Comment by rdhammond** [comment]
Oh. Or it could just tie in directly :)

**Comment by rdhammond** [comment]
I've got it in place, but the disabled color for "Send" may confuse people. Maybe recolor? Maybe add a tooltip to let people know to type something? What do you guys think?

![disabled](https://cloud.githubusercontent.com/assets/801522/6767808/22fb2732-d01d-11e4-94bc-6deeddf98b93.png)

![enabled](https://cloud.githubusercontent.com/assets/801522/6767809/25af2b22-d01d-11e4-88a4-207e0110f521.png)

**Comment by crookedneighbor** [comment]
I think its a pretty normal convention to show disabled buttons with a lighter color scheme. If you wanted to be very clear about it, you could add a tooltip to the send button explaining that you can't send a blank message.

**Comment by rdhammond** [comment]
Oh, yeah, the lighter color thing is definitely convention. Still,
sometimes a user must be hit over the head with it :)

I'll see if I can add a tooltip, then check it in. I assume the linked
issue is handling server-side API code, and it looks like someone's already
done that, but I can look at it too if necessary.

**Comment by crookedneighbor** [comment]
@rdhammond Yeah, just add a tooltip that appears on hover. I think that'll clear up confusion. Anyone that is intentionally trying to send a blank message is going to be confused regardless!

**Comment by rdhammond** [comment]
Here we are. Tooltip only appears when the message box is empty _and_ the button is hovered over. Do you think the copy's okay, or should we change it?

![capture_002_22032015_123412](https://cloud.githubusercontent.com/assets/801522/6770225/0f6aacce-d090-11e4-9b3a-11a156b13f44.png)

**Comment by crookedneighbor** [comment]
I think the text could be better. @lemoness do you have any suggestions?

@rdhammond Is the text hardcoded into the tooltip, or is it a translatable string?

**Comment by rdhammond** [comment]
Translatable.

**Comment by crookedneighbor** [comment]
:+1: 

I'd like to see text that relates more to what's happening, like "Fill in a message" rather than "Type something in the box". Does that make sense? I'd also like to see what @lemoness comes up with.

**Comment by deilann** [comment]
jsyk, I can't see any difference between the two buttons.

**Comment by lemoness** [comment]
I think the tooltip should say "Please type a message."

To make it more obvious, maybe we could also add the same grey box filler
text we have in the Tavern, to say something like "Type your message here."

**Comment by rdhammond** [comment]
I like both! I'll put them in and then submit in a pull request for my
branch.

**Comment by deilann** [comment]
I'm still a little bit concerned about the fact that the active button and
the inactive button look exactly the same to me.

On Mon, Mar 23, 2015 at 9:04 PM, R. Daniel Hammond <notifications@github.com

> wrote:
> 
> I like both! I'll put them in and then submit in a pull request for my
> branch.
>

**Comment by crookedneighbor** [comment]
@deilann the disabled vs not-disabled style issue is a problem with bootstrap. Adding ng-disabled, just adds the disabled attribute to the button, which prevents it from being clicked and it's up to bootstrap to determine how it's styled. 

I think we should open up a separate issue to modify bootstrap a little bit to make it more accessibility friendly.

**Comment by deilann** [comment]
That sounds good to me, Blade!

**Comment by rdhammond** [comment]
Sorry about dragging my feet on this. I should have changes in tonight.

I'm gonna patch up the Meta-Enter issue and try removing the inline-block directive, but I'm ~95% certain the wrapper has to stay. IIRC bootstrap doesn't allow disabled buttons to show tooltips but is smart enough to not show empty tooltips. It's a hack, but I don't see any other (reasonable) way to accomplish it.

**Comment by Alys** [comment]
Never apologise for delays like that. Real life is important too. We're just grateful for any time you can spend on the code!

**Comment by rdhammond** [comment]
Anything else I need to do on this one? It's perfectly fine if it's lower down in the queue priority, I'm just new to using git in team environments and want to make sure I didn't need to resubmit the pull request or something after the latest round of changes.

**Comment by crookedneighbor** [comment]
@rdhammond We get automated emails when new pull requests are made, or when comments are made on a PR. We do not get emails when new changes are pushed to PRs, so if you have something new, and want it reviewed, the best thing to do is to add a comment on the PR to bring it to our attention.

If you're waiting on us for something, and we seem unresponsive, feel free to comment on the PR. We try not to let things slip through the cracks, but sometimes they do.

For our reference, here's the pull request: https://github.com/HabitRPG/habitrpg/pull/4910

**Comment by Alys** [comment]
Fixed in #4910
